### PR TITLE
Problem: versiondb is not ported to main branch

### PR DESCRIPTION
**Pull request by yihuang**
Solution:
- backport the v1.0.4 to main branch, mainly the versiondb feature.

👮🏻👮🏻👮🏻 !!!! REFERENCE THE PROBLEM YOUR ARE SOLVING IN THE PR TITLE AND DESCRIBE YOUR SOLUTION HERE !!!! DO NOT FORGET !!!! 👮🏻👮🏻👮🏻

# PR Checklist:

- [ ] Have you read the [CONTRIBUTING.md](https://github.com/crypto-org-chain/chain-main/blob/master/CONTRIBUTING.md)?
- [ ] Does your PR follow the [C4 patch requirements](https://rfc.zeromq.org/spec:42/C4/#23-patch-requirements)?
- [ ] Have you rebased your work on top of the latest master? 
- [ ] Have you checked your code compiles? (`make`)
- [ ] Have you included tests for any non-trivial functionality?
- [ ] Have you checked your code passes the unit tests? (`make test`)
- [ ] Have you checked your code formatting is correct? (`go fmt`)
- [ ] Have you checked your basic code style is fine? (`golangci-lint run`)
- [ ] If you added any dependencies, have you checked they do not contain any known vulnerabilities? (`go list -json -m all | nancy sleuth`)
- [ ] If your changes affect the client infrastructure, have you run the integration test?
- [ ] If your changes affect public APIs, does your PR follow the [C4 evolution of public contracts](https://rfc.zeromq.org/spec:42/C4/#26-evolution-of-public-contracts)?
- [ ] If your code changes public APIs, have you incremented the crate version numbers and documented your changes in the [CHANGELOG.md](https://github.com/crypto-org-chain/chain-main/blob/master/CHANGELOG.md)?
- [ ] If you are contributing for the first time, please read the agreement in [CONTRIBUTING.md](https://github.com/crypto-org-chain/chain-main/blob/master/CONTRIBUTING.md) now and add a comment to this pull request stating that your PR is in accordance with the [Developer's Certificate of Origin](https://github.com/crypto-org-chain/chain-main/blob/master/CONTRIBUTING.md#developer-certificate-of-originn).

Thank you for your code, it's appreciated! :)

### CODE REVIEW

File: nix/rocksdb.nix [hunk 1]
```diff
@@ -18,13 +18,21 @@
 
 stdenv.mkDerivation rec {
   pname = "rocksdb";
+<<<<<<< HEAD
   version = "7.8.3";
+=======
+  version = "7.9.2";
+>>>>>>> release/v1.0.x
```
Comment (tomtau) [inline nix/rocksdb.nix:25]: the merge conflict should be resolved?

Comment (yihuang) [reply nix/rocksdb.nix]: fixed

File: versiondb/client/changeset.go [hunk 2]
```diff
@@ -0,0 +1,363 @@
...
+	var versionHeader [16]byte
+	if _, err := io.ReadFull(reader, versionHeader[:]); err != nil {
+		return 0, 0, nil, err
+	}
+	version := binary.LittleEndian.Uint64(versionHeader[:8])
+	size := int64(binary.LittleEndian.Uint64(versionHeader[8:16]))
+	var changeSet iavl.ChangeSet
+
+	if size == 0 {
+		return int64(version), 16, &changeSet, nil
```
Comment (tomtau) [inline versiondb/client/changeset.go:83]: the version is unlikely to reach 2^63-1 soon, right?

File: versiondb/client/convert_to_sst.go [hunk 3]
```diff
@@ -0,0 +1,280 @@
...
+	offset += tsrocksdb.TimestampSize
+
+	if pair.Delete {
+		item[offset] = 1
+		offset++
+	} else {
+		copy(item[offset+1:], pair.Value)
+		offset += len(pair.Value) + 1
+	}
+	binary.LittleEndian.PutUint32(item[offset:], uint32(len(pair.Key)))
```
Comment (tomtau) [inline versiondb/client/convert_to_sst.go:216]: is this bound by uint32 -- I saw elsewhere sizes as uint64?

**Review by tomtau — APPROVED** [review]
are all the gosec errors false positives?

File: versiondb/client/changeset.go (hunk 2 above)
Comment (yihuang) [reply versiondb/client/changeset.go:83]: it takes hundreds of years to even overflow uint32 ;D

File: versiondb/client/convert_to_sst.go (hunk 3 above)
Comment (yihuang) [reply versiondb/client/convert_to_sst.go:216]: max key/value length is uint32, both leveldb/rocksdb have that limit, it should be much smaller than that in practice, I'm proposing smaller limit at application level: https://github.com/cosmos/cosmos-sdk/pull/14645

**Comment by yihuang** [comment]
> are all the gosec errors false positives?

there are two complains for `for range` iteration, I'm not sure what's the rationale for that.
all the other complains are for integer type conversion.

File: versiondb/client/convert_to_sst.go (hunk 3 above)
Comment (yihuang) [reply versiondb/client/convert_to_sst.go:216]: > I saw elsewhere sizes as uint64?

The size of an encoded change set of a whole version is encoded as uint64.

**Comment by tomtau** [comment]
there were also file inclusion errors (but that's rather minor to normalise paths) and deferring file.Close()

**Comment by yihuang** [comment]
> there were also file inclusion errors (but that's rather minor to normalise paths) and deferring file.Close()

oh, I see now, didn't click the load more.
file inclusion seems ok, since we don't restrict the directory user can use in the first place, user can put the files anywhere they want.
I think deferring file.Close() is good practice, maybe it means I shouldn't ignore the error return of the Close() call?

**Comment by tomtau** [comment]
yeah, I think it's probably about ignoring potential errors: https://www.joeshaw.org/dont-defer-close-on-writable-files/

**Comment by yihuang** [comment]
> yeah, I think it's probably about ignoring potential errors: https://www.joeshaw.org/dont-defer-close-on-writable-files/

that make sense, I opened an issue for this: https://github.com/crypto-org-chain/cronos/issues/854
this PR is too large for this.

**Comment by mmsqe** [comment]
> > yeah, I think it's probably about ignoring potential errors: https://www.joeshaw.org/dont-defer-close-on-writable-files/
> 
> that make sense, I opened an issue for this: #854 this PR is too large for this.

maybe can return f.Sync() at last, then we can ignore close err

**Comment by yihuang** [comment]
> > > yeah, I think it's probably about ignoring potential errors: https://www.joeshaw.org/dont-defer-close-on-writable-files/
> > 
> > 
> > that make sense, I opened an issue for this: #854 this PR is too large for this.
> 
> maybe can return f.Sync() at last, then we can ignore close err

that's actually the case in the memiavl snapshot writing, maybe applicable to other places too.

**Comment by yihuang** [comment]
the merge queue seems hangs, have no idea how it works 🤦‍♂️

**Comment by tomtau** [comment]
maybe remove and re-add? or just disable merge queues for now?

**Comment by mmsqe** [comment]
maybe should add max retry in test in general, some test stuck for [5h](https://github.com/crypto-org-chain/cronos/actions/runs/4130544275/jobs/7137340203)

File: .github/workflows/build.yml [hunk 4]
```diff
@@ -16,38 +16,63 @@ jobs:
...
+          nix_path: nixpkgs=channel:nixos-22.11
+          extra_nix_config: |
+            access-tokens = github.com=${{ secrets.GITHUB_TOKEN }}
       - uses: actions/setup-go@v3
         with:
           go-version: 1.18
-      - uses: technote-space/get-diff-action@v6
-        id: git_diff
+      - id: changed-files
+        uses: tj-actions/changed-files@v34
```
Comment (julienrbrt) [inline .github/workflows/build.yml:28]: Have you found any advantage or CI speed up using this action?
We still use `get-diff-action` in the SDK, but maybe we should look into this one 👀

Comment (yihuang) [reply .github/workflows/build.yml:28]: We did this switch mainly to handle the github warning of using some deprecated stuff, I think "get-diff-action" is not updated at the time, so we switch to a seemingly better maintained version, which also has rich features.

**Pull request merged**